### PR TITLE
fix: custom tracing in production builds

### PR DESCRIPTION
**Pull request by matthewwalsh0**
## **Description**

Update trace utilities to use Sentry functions via `globalThis` (populated by `setupSentry.js`) to avoid LavaMoat failures in production builds.

Also add trace E2E tests to automatically validate this functionality going forward.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27124?quickstart=1)

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by matthewwalsh0** [comment]
@SocketSecurity ignore-all

File: test/e2e/tests/metrics/traces.spec.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+          .withMetaMetricsController({
+            participateInMetaMetrics: true,
+          })
+          .build(),
+        ganacheOptions: defaultGanacheOptions,
+        title: this.test?.fullTitle(),
+        testSpecificMock: mockSentryCustomTrace,
+        manifestFlags: {
+          doNotForceSentryForThisTest: true,
+        },
```
Comment (HowardBraham) [inline test/e2e/tests/metrics/traces.spec.ts:55]: I commented out all of these
```
manifestFlags: {
 doNotForceSentryForThisTest: true,
},
```
and it still worked, is this expected?  I think that flag is mostly for tests with forced intentional errors.

File: .yarn/patches/lavamoat-core-npm-15.1.1-51fbe39988.patch [hunk 2]
```diff
@@ -1,70 +0,0 @@
...
-+const DIRECT_EVAL_REPLACE_FN = (_, p1) => '(0,eval)' + p1
- function evadeDirectEvalExpressions(source) {
--  /* eslint-disable-next-line prefer-regex-literals */
--  const someDirectEvalPattern = new RegExp('\\beval(\\s*\\()', 'g')
--
--  const replaceFn = (_, p1) => `(0,eval)${p1}`
--  return source.replace(someDirectEvalPattern, replaceFn)
-+  return source.replace(
-+    /\beval(\s*\()/g,
-+    DIRECT_EVAL_REPLACE_FN
```
Comment (naugtur) [inline .yarn/patches/lavamoat-core-npm-15.1.1-51fbe39988.patch:66]: this change is already upstream

File: test/e2e/tests/metrics/traces.spec.ts (hunk 1 above)
Comment (matthewwalsh0) [reply test/e2e/tests/metrics/traces.spec.ts:55]: My reasoning here was a few different things:

- As these tests are verifying Sentry usage itself, it felt odd to use the results for additional metrics.
- When I ran locally with a custom DSN, I got snapshot errors as the host didn't match, this avoids that as it ensures the fake DSN is used.
- There are also negative tests to ensure this doesn't work if metrics are disabled, but I believe `flags.circleci` forces metrics to be enabled unless this manifest flag is set.

**Review by digiwand — DISMISSED** [review]

File: app/scripts/lib/createTracingMiddleware.ts [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Request and repsones are currently untyped.
+// Request and repsonses are currently untyped.
```
Comment (legobeat) [inline app/scripts/lib/createTracingMiddleware.ts:1]:
```suggestion
// Request and responses are currently untyped.
```
:)

**Review by legobeat — APPROVED** [review]
LGTM

**Review by digiwand — APPROVED** [review]

**Review by OGPoyraz — APPROVED** [review]

**Pull request merged**